homeassistant/components/tesla_fleet/climate.py

```diff
@@ -0,0 +1,309 @@
+"""Climate platform for Tesla Fleet integration."""
+
+from __future__ import annotations
+
+from itertools import chain
+from typing import Any, cast
+
+from tesla_fleet_api.const import CabinOverheatProtectionTemp, Scope
+
+from homeassistant.components.climate import (
+    ATTR_HVAC_MODE,
+    ClimateEntity,
+    ClimateEntityFeature,
+    HVACMode,
+)
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    PRECISION_HALVES,
+    PRECISION_WHOLE,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TeslaFleetConfigEntry
+from .const import DOMAIN, TeslaFleetClimateSide
+from .entity import TeslaFleetVehicleEntity
+from .helpers import handle_vehicle_command
+from .models import TeslaFleetVehicleData
+
+DEFAULT_MIN_TEMP = 15
+DEFAULT_MAX_TEMP = 28
+
+PARALLEL_UPDATES = 0
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: TeslaFleetConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Tesla Fleet Climate platform from a config entry."""
+
+    async_add_entities(
+        chain(
+            (
+                TeslaFleetClimateEntity(
+                    vehicle, TeslaFleetClimateSide.DRIVER, entry.runtime_data.scopes
+                )
+                for vehicle in entry.runtime_data.vehicles
+            ),
+            (
+                TeslaFleetCabinOverheatProtectionEntity(
+                    vehicle, entry.runtime_data.scopes
+                )
+                for vehicle in entry.runtime_data.vehicles
+            ),
+        )
+    )
+
+
+class TeslaFleetClimateEntity(TeslaFleetVehicleEntity, ClimateEntity):
+    """Tesla Fleet vehicle climate entity."""
+
+    _attr_precision = PRECISION_HALVES
+
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_hvac_modes = [HVACMode.HEAT_COOL, HVACMode.OFF]
+    _attr_supported_features = (
+        ClimateEntityFeature.TURN_ON
+        | ClimateEntityFeature.TURN_OFF
+        | ClimateEntityFeature.TARGET_TEMPERATURE
+        | ClimateEntityFeature.PRESET_MODE
+    )
+    _attr_preset_modes = ["off", "keep", "dog", "camp"]
+    _enable_turn_on_off_backwards_compatibility = False
+
+    def __init__(
+        self,
+        data: TeslaFleetVehicleData,
+        side: TeslaFleetClimateSide,
+        scopes: Scope,
+    ) -> None:
+        """Initialize the climate."""
+
+        self.read_only = Scope.VEHICLE_CMDS not in scopes or data.signing
+
+        super().__init__(
+            data,
+            side,
+        )
+
+        if self.read_only:
+            self._attr_supported_features = ClimateEntityFeature(0)
+
+    def _async_update_attrs(self) -> None:
+        """Update the attributes of the entity."""
+        value = self.get("climate_state_is_climate_on")
+        if value is None:
```

Comment:
What does this state mean or how does it happen? I am wondering if it should just be `OFF`.  I'm having trouble reasoning about the state where this is none and its readonly below and `self._attr_hvac_modes` is limited and here if its `None` then the hvac modes won't be reset below.

Reply:
When Home Assistant starts up but the vehicle is asleep, most data is no present in the coordinator, hence the key `climate_state_is_climate_on` will not exist and its value will be `None` and it will be shown in Home Assistant as `Unknown` until the vehicle is woken up and the coordinator is refreshed.

Otherwise we only want to check if the value is truthy.

Reply:
so `_attr_hvac_mode` is `None` and what is `_attr_hvac_modes` supposed to be? Can it also be empty?

Reply:
Oh yeah, I guess it can be empty. I have pushed that change and some slight re-ordering of the init to make it work correctly.